core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.seqno;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.IndexSettingsModule;
+import org.junit.Before;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+public class GlobalCheckpointTests extends ESTestCase {
+
+    GlobalCheckpointService checkpointService;
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        checkpointService = new GlobalCheckpointService(new ShardId("test", "_na_", 0),
+            IndexSettingsModule.newIndexSettings("test", Settings.EMPTY));
+    }
+
+    public void testEmptyShards() {
+        assertFalse("no started shards means we should update the checkpoint", checkpointService.updateCheckpointOnPrimary());
+        assertThat(checkpointService.getCheckpoint(), equalTo(SequenceNumbersService.UNASSIGNED_SEQ_NO));
+    }
+
+    public void testGlobalCheckpointUpdate() {
+        Map<String, Long> allocations = new HashMap<>();
+        Set<String> active = new HashSet<>();
+        Set<String> insync = new HashSet<>();
+        Set<String> tracking = new HashSet<>();
+        long maxLocalCheckpoint = Long.MAX_VALUE;
+        for (int i = randomIntBetween(3, 10); i > 0; i--) {
+            String id = "id_" + i + "_" + randomAsciiOfLength(5);
+            long localCheckpoint = randomInt(200);
+            if (randomBoolean()) {
+                active.add(id);
+                maxLocalCheckpoint = Math.min(maxLocalCheckpoint, localCheckpoint);
+            } else if (randomBoolean()) {
+                insync.add(id);
+                maxLocalCheckpoint = Math.min(maxLocalCheckpoint, localCheckpoint);
+            } else {
```

Comment:
Since we go through this loop so few times, do you think it'd be better if we had a uniform distribution over the three cases instead of 0.5/0.25/0.25?